src/emscatter.jl

```diff
@@ -0,0 +1,307 @@
+# ```
+# 
+# After some math, we get: 
+#
+# ```math
+# \alpha_m=\frac{J_m(nkR)J_m(kR)^\prime-\frac{1}{n}J_m(kR)J_m(nkR)^\prime}{\frac{1}{n}H_m^1(kR)J_m(nkr)^\prime-J_m(nkr)H_m^1(kr)^\prime},
+# ```
+# 
+# ```math
+# \beta_m = \frac{H_m^1(kR)J_m(kR)^\prime-J_m(kR)H_m^1(kR)^\prime}{\frac{1}{n}J_m(nkR)^\prime H_m^1(kR)-J_m(nkR)H_m^1(kR)^\prime},
+# ```
+# 
+# where $^\prime$ denotes the derivative, and the derivatives of the Bessel functions are obtained with the recurrent relations: 
+#
+# ```math
+# Y_m(z)^\prime=\frac{Y_{m-1}(z)-Y_{m+1}(z)}{2}
+# ```
+#
+# where $Y_m$ denotes any Bessel functions (Hankel functions).
+#
+# 
+# Finally, the analytical field is ($1/2k$ is the amplitude that comes from the unit line source excitation):
+# ```math
+# H(r>R)=\frac{1}{2k}\sum_m\left[\alpha_mi^mH_m^1(kr)+J_m(kr)\right]e^{im\theta}
+# ```
+#
+# ```math
+# H(r<=R)=\frac{1}{2k}\sum_m\beta_mi^mJ_m(nkr)e^{im\theta}
+# ```
+# 
+
+using SpecialFunctions
+dbesselj(m,z) = (besselj(m-1,z)-besselj(m+1,z))/2
+dhankelh1(m,z)= (hankelh1(m-1,z)-hankelh1(m+1,z))/2
+α(m,n,z) = (besselj(m,n*z)*dbesselj(m,z)-1/n*besselj(m,z)*dbesselj(m,n*z))/(1/n*hankelh1(m,z)*dbesselj(m,n*z)-besselj(m,n*z)*dhankelh1(m,z))
+β(m,n,z) = (hankelh1(m,z)*dbesselj(m,z)-besselj(m,z)*dhankelh1(m,z))/(1/n*dbesselj(m,n*z)*hankelh1(m,z)-besselj(m,n*z)*dhankelh1(m,z))
+
+function H_t(x,xc,r,ϵ,λ)
+    n = √ϵ
+    k = 2*π/λ
+    θ = angle(x[1]-xc[1]+1im*(x[2]-xc[2]))+π
+    M = 20
+    H0 = 0
+    if norm([x[1]-xc[1],x[2]-xc[2]])<=r
+        for m=-M:M
+            H0 += β(m,n,k*r)*cis(m*θ)*besselj(m,n*k*norm([x[1]-xc[1],x[2]-xc[2]]))
+        end
+    else
+        for m=-M:M
+            H0 += α(m,n,k*r)*cis(m*θ)*hankelh1(m,k*norm([x[1]-xc[1],x[2]-xc[2]]))+cis(m*θ)*besselj(m,k*norm([x[1]-xc[1],x[2]-xc[2]]))
+        end
+    end
+    return 1im/(2*k)*H0
+end
+# Construct the analytical solution with finite element functions
+a_theory(u,v) = ∫(u*v)*dΩ
+b_theory(v) = ∫(v*(x->H_t(x,xc,r,ϵ1,λ)))*dΩ
+A_t= assemble_matrix(a_theory,U,V)
+b_t = assemble_vector(b_theory,V)
+u_t = A_t\b_t
+uh_t = FEFunction(U,u_t)
+
+# # Output and compare results
+# 
+# The simulated field is shown below. We can see that the difference between the simulated fields and the analytical solution is almost zero near the cylidner region, which validates the simulation. The difference is larger far away from the cylinder center, that is because we only used M=20 to approximate the sum. As we increase M as well as the resolution, this difference should decrease (until it becomes limited by the PML reflection coefficient from $\sigma_0$ or by floating-point error.)
+# ![](../assets/emscatter/Results.png)
+
+# Save to file and view
+writevtk(Ω,"demo",cellfields=["Real"=>real(uh),
+        "Imag"=>imag(uh),
+        "Norm"=>abs2(uh),
+        "Real_t"=>real(uh_t),
+        "Imag_t"=>imag(uh_t),
+        "Norm_t"=>abs2(uh_t),
+        "Difference"=>abs2(uh_t-uh)])
+#
```

Comment:
Add an `@assert` statement to quantitative check that the solution is correct. This will also allow the CI tests to catch potential regressions in Gridap.

```julia
tol = 10.0e-8
@assert sum(∫(abs2(uh_t-uh))*dΩ) < tol
```

Reply:
(Probably write `tol = 1e-8`  or `tol = 1e-7` … `10e-8 == 1e-7` but is a bit confusing)